### PR TITLE
deps: partial upgrade to hbase 2.5

### PR DESCRIPTION
**Pull request by igorbernstein2**
Unfortunately mirroring-client still has to depend on hbase < 2.4 due to unresolved binary compatibility issues with mutateRow

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java [hunk 1]
```diff
@@ -518,6 +519,48 @@ public void mutateRow(RowMutations rowMutations) throws IOException {
     }
   }
 
+  // Invoked via ByteBuddy
+  // Implements Result mutateRow(RowMutations) for HBase >= 2.4
+  public Result mutateRowResult(RowMutations rowMutations) throws IOException {
+    LOG.trace("mutateRow(RowMutation)");
+    if (rowMutations.getMutations().isEmpty()) {
+      return null;
```
Comment (mutianf) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java:527]: should we throw a invalid argument exception?

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/pom.xml [hunk 2]
```diff
@@ -342,6 +342,13 @@ limitations under the License.
             <version>${hbase2.version}</version>
           </dependency>
 
+          <!-- hbase-shaded-testing-util is missing this dep -->
+          <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>2.28.2</version>
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/pom.xml:349]: should the version be defined somewhere else?

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java [hunk 3]
```diff
@@ -87,6 +96,26 @@ private static <T, R> List<R> map(List<T> list, Function<T, R> f) {
...
+      wrappedReturnType = returnType.getActualTypeArguments()[0];
+    } catch (NoSuchMethodException e) {
+    }
+
+    if (wrappedReturnType == Void.class) {
+      MUTATE_ROW_RETURNS_RESULT = false;
+    } else if (wrappedReturnType == Result.class) {
+      MUTATE_ROW_RETURNS_RESULT = true;
+    } else {
+      MUTATE_ROW_RETURNS_RESULT = null;
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java:115]: what will happen in this case? should we throw exception instead?

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java [hunk 4]
```diff
@@ -316,10 +355,48 @@ public CompletableFuture<Result> increment(Increment increment) {
...
+  // or Append is not present, the return value is now Result.EMPTY
   /** {@inheritDoc} */
   @Override
-  public CompletableFuture<Void> mutateRow(RowMutations rowMutations) {
-    return toCompletableFuture(clientWrapper.mutateRowAsync(hbaseAdapter.adapt(rowMutations)));
+  public CompletableFuture /*<Void|Result>*/ mutateRow(RowMutations rowMutations) {
+    Object emptyReturn = MUTATE_ROW_RETURNS_RESULT ? Result.EMPTY_RESULT : null;
+
+    if (rowMutations.getMutations().isEmpty()) {
+      return CompletableFuture.completedFuture(emptyReturn);
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java:368]: should we throw exception in this case? Or this is acceptable in hbase?

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/org/apache/hadoop/hbase/client/BigtableConnectionRegistry.java [hunk 5]
```diff
@@ -47,6 +47,11 @@ public CompletableFuture<ServerName> getActiveMaster() {
     throw new UnsupportedOperationException();
   }
 
+  @Override
+  public String getConnectionString() {
+    return "fake";
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/org/apache/hadoop/hbase/client/BigtableConnectionRegistry.java:52]: should this be noopString or something?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableTable.java [hunk 6]
```diff
@@ -117,7 +117,13 @@ public void setup() throws IOException {
     when(mockConnection.getBigtableSettings()).thenReturn(settings);
     when(mockBigtableApi.getDataClient()).thenReturn(mockBigtableDataClient);
     when(mockBigtableDataClient.readRows(isA(Query.class))).thenReturn(mockResultScanner);
-    table = new AbstractBigtableTable(mockConnection, hbaseAdapter) {};
+    table =
+        new AbstractBigtableTable(mockConnection, hbaseAdapter) {
+          @Override
+          public void mutateRow(RowMutations rowMutations) throws IOException {
+            mutateRowVoid(rowMutations);
```
Comment (mutianf) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableTable.java:124]: How is table used? does the test need to check the hbase version and call different method?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/org/apache/hadoop/hbase/client/TestAbstractBigtableConnection.java [hunk 7]
```diff
@@ -268,7 +268,12 @@ protected SampledRowKeysAdapter createSampledRowKeysAdapter(
 
     @Override
     public Table getTable(TableName tableName, ExecutorService executorService) {
-      return new AbstractBigtableTable(this, createAdapter(tableName)) {};
+      return new AbstractBigtableTable(this, createAdapter(tableName)) {
+        @Override
+        public void mutateRow(RowMutations rowMutations) throws IOException {
+          mutateRowVoid(rowMutations);
```
Comment (mutianf) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/org/apache/hadoop/hbase/client/TestAbstractBigtableConnection.java:274]: same question here

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java (hunk 1 above)
Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java]: yep empty mutation is ok with empty mutations

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java (hunk 3 above)
Comment (igorbernstein2) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java]: updated it to throw an exception at class instantiation time

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java (hunk 4 above)
Comment (igorbernstein2) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java:377]: no, this is the expected behavior

**Review by mutianf — APPROVED** [review]

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestBasicOps.java [hunk 8]
```diff
@@ -258,18 +258,19 @@ public void testMutateRowCommon() throws IOException {
...
+    long timestamp = 10_000;
     Result result =
         mutateRow(
             table,
             new RowMutations(testRowKey)
                 .add(
                     (Mutation)
                         new Put(testRowKey)
-                            .addColumn(COLUMN_FAMILY, "q".getBytes(), "v1".getBytes()))
+                            .addColumn(COLUMN_FAMILY, "q".getBytes(), timestamp, "v1".getBytes()))
```
Comment (mutianf) [inline bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestBasicOps.java:269]: Why is a timestamp needed now?

**Review by mutianf — APPROVED** [review]

File: bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestBasicOps.java (hunk 8 above)
Comment (igorbernstein2) [reply bigtable-hbase-2.x-parent/bigtable-hbase-2.x-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestBasicOps.java:269]: to allow delete to work later in the test

**Pull request merged**